Add a blog post for visualizing graphs with grafana (#10)
aseriousbiz/blog: _posts/2021/2021-07-01-graphs-in-chat.markdown

---
title: "Visualize Grafana dashboards in chat"
description: "The grafana skill lets you quickly see all your Grafana dashboards and visualize custom graph queries without leaving your chat window."
tags: [grafana,graphs]
excerpt_image: https://user-images.githubusercontent.com/310137/124177982-4369b600-dab1-11eb-84eb-33f3ba1d23c4.png
author:
    avatar: https://gravatar.com/avatar/dc6fa84a4a864f9627e220b716572cba?s=160
    name: shana
---

> An image is worth a thousand words, which is why Abbot can now help you visualize your graphs on chat with the [grafana](https://ab.bot/packages/aseriousbiz/grafana) skill!

Reviewer: ```suggestion
An image is worth a thousand words, which is why [Abbot](https://ab.bot/) can now help you visualize your graphs in chat with the [grafana](https://ab.bot/packages/aseriousbiz/grafana) skill!
```

Author: Oh yes, good point, was going to do that and forgot! :+1:

List all your dashboards and quickly query your panels, generating graphs with custom time ranges and template variables to suit your visualization needs.

## Setting it up
Before you can start visualizing all your graphs, you need to tell Abbot where to go to get them.

> Start by generating a Grafana API key on your Grafana host, usually found at https://your-grafana-host/org/apikeys. Then, visit [https://ab.bot/skills/grafana](https://ab.bot/skills/grafana) and click "Manage skill secrets" to add a secret named ___GrafanaKey___, with the key you created.

Reviewer: I think it looks good as code to indicate they should change it.

```suggestion
Start by generating a Grafana API key on your Grafana host, usually found at `https://your-grafana-host/org/apikeys`. Then, visit [https://ab.bot/skills/grafana](https://ab.bot/skills/grafana) and click "Manage skill secrets" to add a secret named ___GrafanaKey___, with the key you created.
```

```bash
@abbot grafana db "home:memory / cpu"
```

![Screenshot of a panel with a long name](https://user-images.githubusercontent.com/310137/124185001-a7dd4300-daba-11eb-992d-c92e91ec6fd1.png)

You can also use the panel ID instead, so you don't have to type the whole thing every time! Use the `panels` command to see all available panels of a dashboard and get the ID you need:

```bash
@abbot grafana panels home
@abbot grafana db home:11
```

![Screenshot of panels of a dashboard](https://user-images.githubusercontent.com/310137/124184325-b37c3a00-dab9-11eb-883f-3640c0b3a19a.png)

![Screenshot of a graph of a panel by ID](https://user-images.githubusercontent.com/310137/124185258-05718f80-dabb-11eb-8f26-2edfefcfd3e8.png)

## Customize your graphs
Grafana graphs support templated variables and a host of other customization options, and you can use all of those from chat as well - the skill supports custom queries with time ranges, template variables, timezone specifications and size constraints.

Most often, you'll want to customize the time window of your graph. You can specify a range by adding time parameters to your query:

```bash
@abbot grafana db home:3 2d 1d
```

> This generates a graph of the `logins` panel in the `Home` dashboard with a time window of 2 days to 1 day from now.

Reviewer: I'm having trouble parsing what this time window means. Does it mean a time window of 2 days ending 1 day ago? Or a time window starting 2 days ago and ending 1 day ago?

Author: Starting 2 days ago and ending 1 day ago. I'll rephrase!

Time parameters are optional, and if you only have one, it will be interpreted as a window of "now - Time" to now.

```bash
@abbot grafana db "grafana dashboard:client side full page load" 24h var-app=backend var-server=backend_01 var-server=backend_03 var-interval=1h
```

> A more complex query generated a templated Grafana graph with a time window of 24 hours, customized with several variables.

Reviewer: Is this sentence connected to the previous command?

```suggestion
The more complex query above generates a templated Grafana graph with a time window of 24 hours, customized with several variables.
```

![Screenshot showing a graph with custom ranges and template variables](https://user-images.githubusercontent.com/310137/124180188-28e50c00-dab4-11eb-9398-d3333323bae0.png)

> Any parameter that Grafana supports in a query can be passed to the skill as key=value pairs, like timezone specifications or custom graph sizes:

Reviewer: ```suggestion
Any parameter that Grafana supports in a query can be passed to the skill as `key=value` pairs, like timezone specifications or custom graph sizes:
```
